northstar/src/runtime/repository.rs

```diff
@@ -12,100 +12,305 @@
+            return Err(Error::InstallDuplicate(container.clone()));
+        }
         self.containers
-            .insert(container, (dest.to_owned(), npk.manifest().clone()));
+            .insert(container.clone(), (dest.to_owned(), Arc::new(npk)));
 
-        Ok(())
+        info!("Loaded {} into {}", container, self.id);
+
+        Ok(container)
     }
 
-    pub async fn remove(&mut self, container: &Container) -> Result<(), Error> {
-        if let Some((npk, _)) = self.containers.remove(&container) {
-            debug!("Removing {}", npk.display());
-            fs::remove_file(npk)
+    async fn remove(&mut self, container: &Container) -> Result<(), Error>
+    where
+        Self: Sized,
+    {
+        if let Some((path, npk)) = self.containers.remove(&container) {
+            debug!("Removing {} from {}", path.display(), self.id);
+            drop(npk);
+            fs::remove_file(path)
                 .await
-                .map_err(|e| Error::Io("Failed to remove npk".into(), e))
+                .map_err(|e| Error::io("Failed to remove npk", e))
                 .map(drop)
         } else {
-            Err(Error::InvalidContainer(container.clone()))
+            Ok(())
         }
     }
+
+    fn get(&self, container: &Container) -> Option<Arc<Npk>>
+    where
+        Self: Sized,
+    {
+        self.containers.get(container).map(|(_, npk)| npk.clone())
+    }
+
+    fn key(&self) -> Option<&PublicKey>
+    where
+        Self: Sized,
+    {
+        self.key.as_ref()
+    }
+
+    fn containers(&self) -> Vec<Arc<Npk>>
+    where
+        Self: Sized,
+    {
+        self.containers
+            .values()
+            .map(|(_, npk)| npk.clone())
+            .collect()
+    }
+}
+
+/// In memory repository
+#[derive(Default, Debug)]
+pub(super) struct MemRepository {
+    id: RepositoryId,
+    containers: HashMap<Container, (Vec<u8>, Arc<Npk>)>,
+}
+
+impl MemRepository {
+    pub(super) async fn add_buf(&mut self, buf: &[u8]) -> Result<Container, Error> {
+        let data = Vec::from(buf);
+        let fd = Self::memfd_create().map_err(|e| Error::Os("Failed create memfd".into(), e))?;
+        let file = task::block_in_place(|| {
+            let mut file = unsafe { std::fs::File::from_raw_fd(fd) };
+            file.write_all(buf)
+                .map_err(|e| Error::io("Failed copy npk", e))?;
+            file.seek(SeekFrom::Start(0))
+                .map_err(|e| Error::io("Failed seek", e))?;
+            Result::<_, Error>::Ok(BufReader::new(file))
+        })?;
+        debug!("Loading buffer");
+        let npk = npk::Npk::from_reader(file, None).map_err(|e| Error::Npk("Memory".into(), e))?;
+        let manifest = npk.manifest();
+        let container = Container::new(manifest.name.clone(), manifest.version.clone());
+
+        self.containers
+            .insert(container.clone(), (data, Arc::new(npk)));
+
+        info!("Loaded {} into {}", container, self.id);
+
```

Comment:
why not pass in the channel and directly write from it? Then we can get rid of another copy operation also for the in-memory file system.

Reply:
`memfd_create` needs a continuous memory region. Therefore we need to allocate `buf` first. `memfd_create` cannot operate on a stream.